flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java

```diff
@@ -426,30 +425,44 @@ private void commitOperation(
   }
 
   @Override
-  public void processElement(StreamRecord<WriteResult> element) {
-    this.writeResultsOfCurrentCkpt.add(element.getValue());
+  public void processElement(StreamRecord<FlinkWriteResult> element) {
+    FlinkWriteResult flinkWriteResult = element.getValue();
+    List<WriteResult> writeResults =
+        writeResultsSinceLastSnapshot.computeIfAbsent(
+            flinkWriteResult.checkpointId(), k -> Lists.newArrayList());
+    writeResults.add(flinkWriteResult.writeResult());
   }
 
   @Override
   public void endInput() throws IOException {
     // Flush the buffered data files into 'dataFilesPerCheckpoint' firstly.
-    long currentCheckpointId = Long.MAX_VALUE;
-    dataFilesPerCheckpoint.put(currentCheckpointId, writeToManifest(currentCheckpointId));
-    writeResultsOfCurrentCkpt.clear();
-
+    long currentCheckpointId = IcebergStreamWriter.END_INPUT_CHECKPOINT_ID;
+    writeToManifestSinceLastSnapshot(currentCheckpointId);
     commitUpToCheckpoint(dataFilesPerCheckpoint, flinkJobId, operatorUniqueId, currentCheckpointId);
   }
 
+  private void writeToManifestSinceLastSnapshot(long checkpointId) throws IOException {
+    if (!writeResultsSinceLastSnapshot.containsKey(checkpointId)) {
+      dataFilesPerCheckpoint.put(checkpointId, EMPTY_MANIFEST_DATA);
+    }
+
+    for (Map.Entry<Long, List<WriteResult>> writeResultsOfCkpt :
```

Comment:
nit: `writeResultsOfCkpt` -> writeResultsOfCheckpoint

Reply:
should we add Preconditions check that the writeResultsSinceLastSnapshot contains checkpointIds <= the current checkpointId?

Reply:
We may have a problem in case of concurrent checkpoints (say 1 and 2). when checkpoint 1 snapshot method is executed, we shall not write the data files partially received for checkpoint 2 to manifest file. we should only process the map entries up to the checkpointId

Reply:
> should we add Preconditions check that the writeResultsSinceLastSnapshot contains checkpointIds <= the current checkpointId?

I think we can't add a Preconditions check, because the method needs to consider the case where writeResultsSinceLastSnapshot is empty.

> We may have a problem in case of concurrent checkpoints (say 1 and 2). when checkpoint 1 snapshot method is executed, we shall not write the data files partially received for checkpoint 2 to manifest file. we should only process the map entries up to the checkpointId

I discussed this scenario (If I understand correctly) with @pvary and we concluded that it is not possible. https://github.com/apache/iceberg/pull/10526#discussion_r1650849668

Reply:
Checked out the linked comment. It is the same scenario. But I think it could happen. let me call out the step where it can be problematic when concurrent checkpoints are allowed. cc @pvary 

- Writer 1 writes data
- Writer 2 writes data
- Writer 1 prepareSnapshotPreBarrier for Checkpoint 1, File 1-1 with checkpoint 1 created
- Committer receives File 1-1
- Writer 2 prepareSnapshotPreBarrier for Checkpoint 1, File 1-2 with checkpoint 1 created
- Committer receives File 1-2
- Writer 1 writes data
- Writer 2 writes data
- Writer 1 prepareSnapshotPreBarrier for Checkpoint 2, File 2-1 with checkpoint 2 created
- Committer receives File 2-1
- Committer starts snapshotState for Checkpoint 1 - At this stage we have writeResultsSinceLastSnapshot with more data, but not all of them are ready to commit.  `<--- In this step, current implementation will commit files from checkpoint 2, which is incorrect. commuter should only commit files up to checkpoint 1.`
- Committer receives File 2-2
- Committer starts snapshotState for Checkpoint 2

Reply:
@stevenzwu Thank you for the detailed description.

I verified the above scenario through the program, refer to : https://github.com/zhongqishang/flink-iceberg-sink-mock/blob/master/src/main/java/org/example/DataStreamJob.java.

Step 10 and step 11 cannot happen in the above order.

Besides, if there is this order, then the current implementation (before this PR)is also problematic.

Reply:
I think the process above will translate to the following messages on the wire:
- Writer-1 prepareSnapshotBarrier will emit File 1-1 generated by the method
- Writer-1 will emits the checkpoint barrier for CHK-1
- Writer-2 prepareSnapshotBarrier will emit File 1-2 generated by the method
- Writer-2 will emits checkpoint barrier for CHK-1
- Committer receives File 1-1
- Committer receives File 1-2
- Writer-1 prepareSnapshotBarrier will emit File 2-1 generated by the method
- Writer-1 will emits checkpoint barrier for CHK-2
- Committer will **NOT** receive File 2-1 until it received the checkpoint barrier for every upstream operator, and executed the snapshotState method.

Based on this the fix should be working if aligned checkpoints are used. I think all bets are out of the window, when unaligned checkpoints are used, as this case the checkpoint barriers can "overtake" normal records.

@gyfora: Could you please validate if my understanding above is correct?

If I am right, it would be good to state this somewhere in the Iceberg Flink connector documentation.

Reply:
@pvary @stevenzwu @zhongqishang The checkpointing algorithm guarantees that when snapshot method is called on an operator that operator cannot see any data/records from later checkpoints. This applies for both aligned/unaligned checkpoints.

So the committer never receives files from CHK 2 before snapshotting/committing CHK 1 (regardless of aligned/unaligned case)